chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,177 @@
+        S3 is relatively cheap, as you only pay for what you use, in addition to it being managed by Amazon which have the resources and the knowledge to provide extremely high availability (11 nines) with good performance and a good cost.
+        S3 also uses Erasure coding to have a smaller replication overhead so it allows for redundency without too large of an overhead.
+        S3 requires also higher on prem costs, because the recomended S3 on prem hardware reqs are a NVME drive (which is very fast but also very very expensive), along with 256 gb of ram and 100 GBE network, because 
+        So on prem it is slightly more expensive than hadoop.
+
+- [x] *Hdfs use cases*: to store large amounts of files that need to be searchable (Websites and text / pdf files), hadoop is often used in the finance industry for risk modelling (using big data that is stored in hdfs and analyzing with mapreduce and spark), can also be used like a regular file / block storage that can handle a lot of data and scale easily while being reliable, Hdfs can also be used along with analytics tools to detect fradulent activity.
+
+- [x] *Terms*:
+  - [x] *Failover*: The process of handling a node / master failure, when a leader fails, a failover event is triggered and a leader re election starts.
+  - [x] *Fencing*: Isolation of a node or shared resources when a node seems to be malfunctioning.
+  - [x] *Zkfc*:  The Zookeeper Failover Controller which is responsible for automatic failover in distributed systems like Hadoop (namenode failover) and Kafka (Controller failover)
+  - [x] *Quorum*: Quorum (in zookeeper) is the smallest amount of running servers of a distributed system in order for it to work and be available for the clients, usually the formula for a quorum is that if you have 2n + 1 nodes, n nodes can fail and the system will keep working as it worked before.
+  The quorum can also be used to elect leaders, by having a quorum of the in sync nodes that can become leaders in the case of node failure.
+
+
 1. **Q:** What is Apache Hadoop?
+1. **A:** Hadoop is a distributed data processing framework that works based on a master-slave architecture which was built with the assumption that hardware failure is the norm and allows for big data processing with easy scaling, to thousands of nodes.
+Hadoop has a few components in it that allow for optimal resource usage and job schedualing while allowing multiple jobs simountaneusly, Hadoop also includes Hdfs which is a distributed file system that saves tha
+Hadoop also contains YARN which allows for better and more efficient resource utilization in addition to allowing multiple processes to run in parrallel.
```

Comment:
What other components does hadoop contain? no need to go into detail, just what their job is in hadoop